mdata/write_buffer.go

```diff
@@ -0,0 +1,84 @@
+package mdata
+
+import (
+	"gopkg.in/raintank/schema.v1"
+)
+
+/*
+ * The write buffer keeps a window of data during which it is ok to send data out of order.
+ * Once the reorder window has passed it will try to flush the data out.
+ * The write buffer itself is not thread safe because it is only used by AggMetric,
+ * which is thread safe, so there is no locking in the buffer.
+ */
+
+type ReorderBuffer struct {
+	len      uint32                // length of buffer in number of datapoints
+	newest   uint32                // index of newest buffer entry
+	interval uint32                // metric interval
+	buf      []schema.Point        // the actual buffer holding the data
+	flush    func(uint32, float64) // flushCount callback
```

Comment:
i'm not sure exactly what is the cost of having a value of type func (remember, multiplied for every single metric that has a reorderbuffer), but it's gotta be at least a few bytes. probably internally a pointer too, causing some GC overhead.

my suggestion is to remove this callback rather `Add` can return a slice of points, so that the caller (AggMetric) can just add all the points that it gets back from the RB.  This way `Add` becomes more of a "Process". a point goes in, and 0 or more come out.